{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "significant-disposal",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "What crop are you going to input?\n",
      "         Choose one of the following:\n",
      "         Ca/To/Ke/Te/Me/Ja/Bm/Ka/Se/Kp/Cs\n",
      "/mnt/d/Genotype Result 2/Capsicum_annuum/Pltd.snp.xls\n",
      "What crop are you going to input?\n",
      "             Choose one of the following:\n",
      "             Ca/To/Ke/Te/Me/Ja/Bm/Ka/Se/Kp/Cs\n",
      "Ca\n",
      "Are you working with single input file or multiple input files? Choose [S/M] \n",
      " S\n",
      "Enter file path: /mnt/d/Genotype Result 2/Capsicum_annuum/Pltd.snp.xls\n",
      "Working with /mnt/d/Genotype Result 2/Capsicum_annuum/Pltd.snp.xls file\n",
      "Now you have Ca_snp-001_SNP_Information.csv file\n",
      "Now you have Ca_snp-001_Sequence_data.csv file\n",
      "Now you have Ca_snp-001_Sample_genotype.csv file\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "This software serves to prepare the input files for Genomic database.\n",
    "The expected input file is the sequencing report from BGI,\n",
    "where the header consists of #Chr,Loci, REF, series of sample name, and Sequencing Details.\n",
    "This script partitions the original file into three csv files\n",
    "following the entities format of the Genomic database.\n",
    "\n",
    "Developed by Sausan Nafisah | September 2021\n",
    "'''\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "#_ordering sample    \n",
    "def atoi(text):\n",
    "    return int(text) if text.isdigit() else text\n",
    "\n",
    "def natural_keys(text):\n",
    "    return [ atoi(c) for c in re.split('(\\d+)',text) ]\n",
    "\n",
    "#_main program\n",
    "def genomic_db_data_preparation_software(file):\n",
    "    global crop\n",
    "    #data manipulation\n",
    "    df = pd.read_csv(file, sep = '\\t')\n",
    "\n",
    "    Chr = df['#Chr'].tolist()\n",
    "    Loci = df['Loci'].tolist()\n",
    "    Ref = df['REF'].tolist()\n",
    "    Sequence_detail = df['Sequence_Detail'].tolist()\n",
    "\n",
    "    #__create index\n",
    "    row, cols = df.shape\n",
    "    index = []\n",
    "    location = Chr[0]\n",
    "      \n",
    "    for i in range(row):\n",
    "        item = '{0}_{1}_{2}'.format(crop, location, Loci[i])\n",
    "        index.append(item)\n",
    "\n",
    "    #__check point\n",
    "    #print('index [600]: ',index[600])\n",
    "\n",
    "    #__Obtaining the sample name\n",
    "    column_name = df.columns.tolist()\n",
    "\n",
    "    moved = ['#Chr','Loci', 'REF', 'Sequence_Detail']\n",
    "\n",
    "    for item in moved:\n",
    "        #print('removing {} from the dataset'.format(item))\n",
    "        column_name.remove(item)\n",
    "        \n",
    "    #__ordering sample \n",
    "    column_name.sort(key=natural_keys)\n",
    "    #__Check point\n",
    "    #print (column_name)\n",
    "\n",
    "    #__preparing genotype dataset\n",
    "    genotype = {}\n",
    "    for column in column_name:\n",
    "        genotype1 = {'{}'.format(column):df['{}'.format(column)].tolist()}\n",
    "        genotype.update(genotype1)\n",
    "        \n",
    "    #__preparing input file\n",
    "    SNP_Information = {'rsID':index, 'Chr':Chr, 'Loci':Loci, 'Ref':Ref}\n",
    "    Sequence_data = {'rsID':index, 'Sequencing_detail':Sequence_detail}\n",
    "    Sample_Genotype = {'rsID':index}\n",
    "    Sample_Genotype.update(genotype)\n",
    "    \n",
    "    input_file1 = pd.DataFrame(SNP_Information)\n",
    "    input_file2 = pd.DataFrame(Sequence_data)\n",
    "    input_file3 = pd.DataFrame(Sample_Genotype)\n",
    "    \n",
    "    return input_file1, input_file2, input_file3\n",
    "\n",
    "#_Working with single input file\n",
    "def single_file():\n",
    "    file = input('Enter file path: ')\n",
    "    \n",
    "    print(\"Working with {} file\".format(file))\n",
    "    input_file1, input_file2, input_file3 = genomic_db_data_preparation_software(file)\n",
    "    \n",
    "    input_file1.to_csv('{0}_snp-001_SNP_Information.csv'.format(crop), index=False)\n",
    "    print(\"Now you have {0}_snp-001_SNP_Information.csv file\".format(crop))\n",
    "    \n",
    "    input_file2.to_csv('{0}_snp-001_Sequence_data.csv'.format(crop), index=False)\n",
    "    print(\"Now you have {0}_snp-001_Sequence_data.csv file\".format(crop))\n",
    "    \n",
    "    input_file3.to_csv('{0}_snp-001_Sample_genotype.csv'.format(crop), index=False)\n",
    "    print(\"Now you have {0}_snp-001_Sample_genotype.csv file\".format(crop))\n",
    "    \n",
    "#_Working with multiple input files\n",
    "def multi_files():\n",
    "    file_list = input('Enter the path to the file containing the file list: ')\n",
    "    \n",
    "    for line in open(file_list):\n",
    "        edit_line = line.split('\\n')\n",
    "        print(\"Working with {} file\".format(edit_line[0]))\n",
    "        file = '/mnt/d/Genotype Result 2/Solanum_lycopersicum/{}'.format(edit_line[0])\n",
    "        input_file1, input_file2, input_file3 = genomic_db_data_preparation_software(file)\n",
    "        \n",
    "        #_Creating input files\n",
    "        file_id = edit_line[0].split('.')\n",
    "        input_file1.to_csv('{0}_{1}_SNP_Information.csv'.format(crop, file_id[0]), index=False)\n",
    "        print(\"Now you have {0}_{1}_SNP_Information.csv file\".format(crop, file_id[0]))\n",
    "        \n",
    "        input_file2.to_csv('{0}_{1}_Sequence_data.csv'.format(crop, file_id[0]), index=False)\n",
    "        print('Now you have {0}_{1}_Sequence_data.csv file'.format(crop, file_id[0]))\n",
    "        \n",
    "        input_file3.to_csv('{0}_{1}_Sample_genotype.csv'.format(crop, file_id[0]), index=False)\n",
    "        print('Now you have {0}_{1}_Sample_genotype.csv file'.format(crop, file_id[0]))\n",
    "        print('\\n')\n",
    "        \n",
    "if __name__ == \"__main__\":\n",
    "    #input crop info\n",
    "    global crop\n",
    "    crop = input('What crop are you going to input?\\n \\\n",
    "        Choose one of the following:\\n \\\n",
    "        Ca/To/Ke/Te/Me/Ja/Bm/Ka/Se/Kp/Cs\\n')\n",
    "    \n",
    "    crop_list = ['Ca','To','Ke','Te','Me','Ja','Bm','Ka','Se','Kp','Cs']\n",
    "    \n",
    "    while crop not in crop_list:\n",
    "        crop = input('What crop are you going to input?\\n \\\n",
    "            Choose one of the following:\\n \\\n",
    "            Ca/To/Ke/Te/Me/Ja/Bm/Ka/Se/Kp/Cs\\n')\n",
    "    \n",
    "    #choose working type\n",
    "    working_type = input('Are you working with single input file or multiple input files? Choose [S/M] \\n ')\n",
    "    \n",
    "    if working_type == 'S':\n",
    "        single_file()\n",
    "    elif working_type == 'M':\n",
    "        multi_files()\n",
    "    else:\n",
    "        print('you typed a wrong answer.')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
