{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "13378d85",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter file path or file name: /mnt/c/Users/sausan.nafisah/Downloads/Pltd.snp.xls\n",
      "What crop are you going to input?\n",
      "         Choose one of the following:\n",
      "         Ca/To/Ke/Te/Me/Ja/Bm/Ka/Se/Kp/Cs\n",
      "Ca\n",
      "Now you have Ca_SNP_Information.csv file\n",
      "Now you have Ca_Sequence_data.csv file\n",
      "Now you have Ca_Sample_genotype.csv file\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "#file = '/mnt/d/Genotype Result 2/Capsicum_annuum/Pltd.snp.xls'\n",
    "#file = '/mnt/c/Users/sausan.nafisah/Downloads/Pltd.snp.xls'\n",
    "\n",
    "#ordering sample    \n",
    "def atoi(text):\n",
    "    return int(text) if text.isdigit() else text\n",
    "\n",
    "def natural_keys(text):\n",
    "    return [ atoi(c) for c in re.split('(\\d+)',text) ]\n",
    "\n",
    "#main program\n",
    "def genomic_db_data_preparation_software(file):\n",
    "    #data manipulation\n",
    "    df = pd.read_csv(file, sep = '\\t')\n",
    "\n",
    "    Chr = df['#Chr'].tolist()\n",
    "    Loci = df['Loci'].tolist()\n",
    "    Ref = df['REF'].tolist()\n",
    "    Sequence_detail = df['Sequence_Detail'].tolist()\n",
    "\n",
    "    #create index\n",
    "    row, cols = df.shape\n",
    "    index = []\n",
    "    location = Chr[0]\n",
    "    crop_list = ['Ca','To','Ke','Te','Me','Ja','Bm','Ka','Se','Kp','Cs']\n",
    "    crop = ''\n",
    "\n",
    "    while crop not in crop_list:\n",
    "        crop = input('What crop are you going to input?\\n \\\n",
    "        Choose one of the following:\\n \\\n",
    "        Ca/To/Ke/Te/Me/Ja/Bm/Ka/Se/Kp/Cs\\n')\n",
    "\n",
    "    for i in range(row):\n",
    "        item = '{0}_{1}_{2}'.format(crop, location, Loci[i])\n",
    "        index.append(item)\n",
    "\n",
    "    #check point\n",
    "    #print('index [600]: ',index[600])\n",
    "\n",
    "    #Obtaining the sample name\n",
    "    column_name = df.columns.tolist()\n",
    "\n",
    "    moved = ['#Chr','Loci', 'REF', 'Sequence_Detail']\n",
    "\n",
    "    for item in moved:\n",
    "        #print('removing {} from the dataset'.format(item))\n",
    "        column_name.remove(item)\n",
    "        \n",
    "    #ordering sample \n",
    "    column_name.sort(key=natural_keys)\n",
    "    #Check point\n",
    "    #print (column_name)\n",
    "\n",
    "    #preparing genotype dataset\n",
    "    genotype = {}\n",
    "    for column in column_name:\n",
    "        #selection = '{}'.format(column)\n",
    "        genotype1 = {'{}'.format(column):df['{}'.format(column)].tolist()}\n",
    "        genotype.update(genotype1)\n",
    "        \n",
    "    #preparing csv file as Genomic db input file\n",
    "    SNP_Information = {'rsID':index, 'Chr':Chr, 'Loci':Loci, 'Ref':Ref}\n",
    "    Sequence_data = {'rsID':index, 'Sequencing_detail':Sequence_detail}\n",
    "    Sample_Genotype = {'rsID':index}\n",
    "    Sample_Genotype.update(genotype)\n",
    "    \n",
    "    #creating input files\n",
    "    input_file1 = pd.DataFrame(SNP_Information)\n",
    "    input_file1.to_csv('{}_SNP_Information.csv'.format(crop), index=False)\n",
    "    print(\"Now you have {}_SNP_Information.csv file\".format(crop))\n",
    "    \n",
    "    input_file2 = pd.DataFrame(Sequence_data)\n",
    "    input_file2.to_csv('{}_Sequence_data.csv'.format(crop), index=False)\n",
    "    print('Now you have {}_Sequence_data.csv file'.format(crop))\n",
    "    \n",
    "    input_file3 = pd.DataFrame(Sample_Genotype)\n",
    "    input_file3.to_csv('{}_Sample_genotype.csv'.format(crop), index=False)\n",
    "    print('Now you have {}_Sample_genotype.csv file'.format(crop))\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    file = input('Enter file path or file name: ')\n",
    "    genomic_db_data_preparation_software(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "555770e9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
