{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "significant-disposal",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working with Chr01.snp.xls file\n",
      "Now you have Ca_Chr01_SNP_Information.csv file\n",
      "Now you have Ca_Chr01_Sequence_data.csv file\n",
      "Now you have Ca_Chr01_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with Chr02.snp.xls file\n",
      "Now you have Ca_Chr02_SNP_Information.csv file\n",
      "Now you have Ca_Chr02_Sequence_data.csv file\n",
      "Now you have Ca_Chr02_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with Chr03.snp.xls file\n",
      "Now you have Ca_Chr03_SNP_Information.csv file\n",
      "Now you have Ca_Chr03_Sequence_data.csv file\n",
      "Now you have Ca_Chr03_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with Chr04.snp.xls file\n",
      "Now you have Ca_Chr04_SNP_Information.csv file\n",
      "Now you have Ca_Chr04_Sequence_data.csv file\n",
      "Now you have Ca_Chr04_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with Chr05.snp.xls file\n",
      "Now you have Ca_Chr05_SNP_Information.csv file\n",
      "Now you have Ca_Chr05_Sequence_data.csv file\n",
      "Now you have Ca_Chr05_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with Chr06.snp.xls file\n",
      "Now you have Ca_Chr06_SNP_Information.csv file\n",
      "Now you have Ca_Chr06_Sequence_data.csv file\n",
      "Now you have Ca_Chr06_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with Chr07.snp.xls file\n",
      "Now you have Ca_Chr07_SNP_Information.csv file\n",
      "Now you have Ca_Chr07_Sequence_data.csv file\n",
      "Now you have Ca_Chr07_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with Chr08.snp.xls file\n",
      "Now you have Ca_Chr08_SNP_Information.csv file\n",
      "Now you have Ca_Chr08_Sequence_data.csv file\n",
      "Now you have Ca_Chr08_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with Chr09.snp.xls file\n",
      "Now you have Ca_Chr09_SNP_Information.csv file\n",
      "Now you have Ca_Chr09_Sequence_data.csv file\n",
      "Now you have Ca_Chr09_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with Chr10.snp.xls file\n",
      "Now you have Ca_Chr10_SNP_Information.csv file\n",
      "Now you have Ca_Chr10_Sequence_data.csv file\n",
      "Now you have Ca_Chr10_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with Chr11.snp.xls file\n",
      "Now you have Ca_Chr11_SNP_Information.csv file\n",
      "Now you have Ca_Chr11_Sequence_data.csv file\n",
      "Now you have Ca_Chr11_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with Chr12.snp.xls file\n",
      "Now you have Ca_Chr12_SNP_Information.csv file\n",
      "Now you have Ca_Chr12_Sequence_data.csv file\n",
      "Now you have Ca_Chr12_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with MT.snp.xls file\n",
      "Now you have Ca_MT_SNP_Information.csv file\n",
      "Now you have Ca_MT_Sequence_data.csv file\n",
      "Now you have Ca_MT_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with Other.snp.xls file\n",
      "Now you have Ca_Other_SNP_Information.csv file\n",
      "Now you have Ca_Other_Sequence_data.csv file\n",
      "Now you have Ca_Other_Sample_genotype.csv file\n",
      "\n",
      "\n",
      "Working with Pltd.snp.xls file\n",
      "Now you have Ca_Pltd_SNP_Information.csv file\n",
      "Now you have Ca_Pltd_Sequence_data.csv file\n",
      "Now you have Ca_Pltd_Sample_genotype.csv file\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "This software serves to prepare the input files for Genomic database.\n",
    "The expected input file is the sequencing report from BGI,\n",
    "where the header consists of #Chr,Loci, REF, series of sample name, and Sequencing Details.\n",
    "This script partitions the original file into three csv files\n",
    "following the entities format of the Genomic database.\n",
    "\n",
    "Developed by Sausan Nafisah | September 2021\n",
    "'''\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "#file = '/mnt/d/Genotype Result 2/Capsicum_annuum/Chr01.snp.xls'\n",
    "#file = '/mnt/d/Genotype Result 2/Cucumis_sativus/combine.snp-001.xls'\n",
    "\n",
    "#ordering sample    \n",
    "def atoi(text):\n",
    "    return int(text) if text.isdigit() else text\n",
    "\n",
    "def natural_keys(text):\n",
    "    return [ atoi(c) for c in re.split('(\\d+)',text) ]\n",
    "\n",
    "#main program\n",
    "def genomic_db_data_preparation_software(file):\n",
    "    #data manipulation\n",
    "    df = pd.read_csv(file, sep = '\\t')\n",
    "\n",
    "    Chr = df['#Chr'].tolist()\n",
    "    Loci = df['Loci'].tolist()\n",
    "    Ref = df['REF'].tolist()\n",
    "    Sequence_detail = df['Sequence_Detail'].tolist()\n",
    "\n",
    "    #create index\n",
    "    row, cols = df.shape\n",
    "    index = []\n",
    "    location = Chr[0]\n",
    "    crop_list = ['Ca','To','Ke','Te','Me','Ja','Bm','Ka','Se','Kp','Cs']\n",
    "    crop = 'Ca'\n",
    "\n",
    "    while crop not in crop_list:\n",
    "        crop = input('What crop are you going to input?\\n \\\n",
    "        Choose one of the following:\\n \\\n",
    "        Ca/To/Ke/Te/Me/Ja/Bm/Ka/Se/Kp/Cs\\n')\n",
    "\n",
    "    for i in range(row):\n",
    "        item = '{0}_{1}_{2}'.format(crop, location, Loci[i])\n",
    "        index.append(item)\n",
    "\n",
    "    #check point\n",
    "    #print('index [600]: ',index[600])\n",
    "\n",
    "    #Obtaining the sample name\n",
    "    column_name = df.columns.tolist()\n",
    "\n",
    "    moved = ['#Chr','Loci', 'REF', 'Sequence_Detail']\n",
    "\n",
    "    for item in moved:\n",
    "        #print('removing {} from the dataset'.format(item))\n",
    "        column_name.remove(item)\n",
    "        \n",
    "    #ordering sample \n",
    "    column_name.sort(key=natural_keys)\n",
    "    #Check point\n",
    "    #print (column_name)\n",
    "\n",
    "    #preparing genotype dataset\n",
    "    genotype = {}\n",
    "    for column in column_name:\n",
    "        #selection = '{}'.format(column)\n",
    "        genotype1 = {'{}'.format(column):df['{}'.format(column)].tolist()}\n",
    "        genotype.update(genotype1)\n",
    "        \n",
    "    #preparing input file\n",
    "    SNP_Information = {'rsID':index, 'Chr':Chr, 'Loci':Loci, 'Ref':Ref}\n",
    "    Sequence_data = {'rsID':index, 'Sequencing_detail':Sequence_detail}\n",
    "    Sample_Genotype = {'rsID':index}\n",
    "    Sample_Genotype.update(genotype)\n",
    "    \n",
    "    input_file1 = pd.DataFrame(SNP_Information)\n",
    "    input_file2 = pd.DataFrame(Sequence_data)\n",
    "    input_file3 = pd.DataFrame(Sample_Genotype)\n",
    "    \n",
    "    return input_file1, input_file2, input_file3\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    file_list = '/mnt/c/Users/sausan.nafisah/Documents/list_data_genotype.txt'\n",
    "    crop = 'Ca'\n",
    "    for line in open(file_list):\n",
    "        edit_line = line.split('\\n')\n",
    "        print(\"Working with {} file\".format(edit_line[0]))\n",
    "        file = '/mnt/d/Genotype Result 2/Capsicum_annuum/{}'.format(edit_line[0])\n",
    "        input_file1, input_file2, input_file3 = genomic_db_data_preparation_software(file)\n",
    "        \n",
    "        #Creating input files\n",
    "        file_id = edit_line[0].split('.')\n",
    "        input_file1.to_csv('{0}_{1}_SNP_Information.csv'.format(crop, file_id[0]), index=False)\n",
    "        print(\"Now you have {0}_{1}_SNP_Information.csv file\".format(crop, file_id[0]))\n",
    "        \n",
    "        input_file2.to_csv('{0}_{1}_Sequence_data.csv'.format(crop, file_id[0]), index=False)\n",
    "        print('Now you have {0}_{1}_Sequence_data.csv file'.format(crop, file_id[0]))\n",
    "        \n",
    "        input_file2.to_csv('{0}_{1}_Sample_genotype.csv'.format(crop, file_id[0]), index=False)\n",
    "        print('Now you have {0}_{1}_Sample_genotype.csv file'.format(crop, file_id[0]))\n",
    "        print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "coastal-bennett",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Loci</th>\n",
       "      <th>Ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ca_Chr10_116</td>\n",
       "      <td>Chr10</td>\n",
       "      <td>116</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ca_Chr10_136</td>\n",
       "      <td>Chr10</td>\n",
       "      <td>136</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ca_Chr10_169</td>\n",
       "      <td>Chr10</td>\n",
       "      <td>169</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ca_Chr10_181</td>\n",
       "      <td>Chr10</td>\n",
       "      <td>181</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ca_Chr10_226</td>\n",
       "      <td>Chr10</td>\n",
       "      <td>226</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           rsID    Chr  Loci Ref\n",
       "0  Ca_Chr10_116  Chr10   116   T\n",
       "1  Ca_Chr10_136  Chr10   136   C\n",
       "2  Ca_Chr10_169  Chr10   169   A\n",
       "3  Ca_Chr10_181  Chr10   181   G\n",
       "4  Ca_Chr10_226  Chr10   226   G"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_cabai = pd.read_csv('Ca_Chr10_SNP_Information.csv')\n",
    "data_cabai.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "interested-appointment",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
